### ISSUE TITLE
Add context manager to temporarily disable GC

### ISSUE THREAD

**Issue by rhettinger** [python/cpython#75537]
BPO | [31356](https://bugs.python.org/issue31356)
--- | :---
Nosy | @rhettinger, @gpshead, @ncoghlan, @pitrou, @ned-deily, @pmp-p, @ericsnowcurrently, @serhiy-storchaka, @1st1, @MojoVampire, @YoSTEALTH, @bharel, @lisroach, @pablogsal, @sweeneyde, @Jongy
PRs | <li>python/cpython#4224</li><li>python/cpython#5456</li><li>python/cpython#5462</li><li>python/cpython#5495</li><li>python/cpython#5496</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2017-09-05.23:47:23.956>
labels = ['interpreter-core', '3.8', 'type-feature', 'library']
title = 'Add context manager to temporarily disable GC'
updated_at = <Date 2021-05-01.15:07:54.862>
user = 'https://github.com/rhettinger'
```

bugs.python.org fields:
```python
activity = <Date 2021-05-01.15:07:54.862>
actor = 'Yonatan Goldschmidt'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Interpreter Core', 'Library (Lib)']
creation = <Date 2017-09-05.23:47:23.956>
creator = 'rhettinger'
dependencies = []
files = []
hgrepos = []
issue_num = 31356
keywords = ['patch']
message_count = 52.0
messages = ['301407', '301633', '301636', '301644', '301645', '301710', '301720', '302903', '305402', '306053', '306054', '306071', '306072', '306076', '311167', '311300', '311305', '311312', '311331', '311337', '311340', '311342', '311343', '311344', '311347', '311348', '311350', '311351', '311352', '311356', '311357', '311360', '311362', '311363', '311366', '311372', '311396', '311397', '311402', '311444', '311491', '311494', '311495', '311508', '311512', '311513', '311514', '311566', '311635', '311646', '385213', '385214']
nosy_count = 16.0
nosy_names = ['rhettinger', 'gregory.p.smith', 'ncoghlan', 'pitrou', 'ned.deily', 'pmpp', 'eric.snow', 'serhiy.storchaka', 'yselivanov', 'josh.r', 'YoSTEALTH', 'bar.harel', 'lisroach', 'pablogsal', 'Dennis Sweeney', 'Yonatan Goldschmidt']
pr_nums = ['4224', '5456', '5462', '5495', '5496']
priority = 'normal'
resolution = None
stage = 'needs patch'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue31356'
versions = ['Python 3.8']
```

</p></details>

**Comment by rhettinger** [comment]
Used like this:

```py
    with gc_disabled():
        run_some_timing()

    with gc_disabled():
        # do_something_that_has_real_time_guarantees
        # such as a pair trade, robotic braking, etc
```


This would be implemented in C for atomicity and speed.  It would be roughly equivalent to:

```py
    @contextmanager
    def gc_disabled():
        if gc.isenabled():
            gc.disable()
            yield
            gc.enable()
        else:
            yield
```

**Comment by ncoghlan** [comment]
+1 from me for the general idea. As far as where to put it goes, I think the `gc` module would be the most appropriate home.

**Comment by serhiy-storchaka** [comment]
Note that threads can break this. Even without calling gc.enable() explicitly, "with gc_disabled()" in different thread can enable GC inside other "with gc_disabled()" block.

**Comment by ncoghlan** [comment]
Yes, this will be in the same category as the standard stream redirection context managers - multi-threaded applications either won't be able to use it, or else will need to manage access to it somehow.

**Comment by gpshead** [comment]
I believe Raymond is aware of the thread issue.  We discussed it.

If gc.disable() would return the previous state of the gc instead of None and an API to enable based on a passed in bool, both of which are written in C and executed with the GIL (or merely another lock protecting changing the gc state rather than the GIL) held and this Python would works with multiple threads all fighting to control the gc state:

```py
@contextmanager
def gc_disabled():
  previous_enable_state = gc.disable()
  yield
  gc.set_enable(previous_enable_state)
```


But we don't need to modify gc.disable's return value or add a set_enable() if gc_disabled() itself is not implemented in Python.  That's up to the implementer.

**Comment by 99ffcaa5-b43b-4e8e-a35e-9c890007b9cd** [comment]
I'd be -1 on this without a demonstrated broad need for this in at least some context outside of microbenchmarking utilities (which presumably have already implemented similar stuff).

If a minimum bar for applicability isn't applied, we'll end up with dozens of these special purpose managers justified by the last limited applicability one. Stuff like contextlib.closing is justifiable (though I wish it allowed you to replace the method name called, so it could call terminate, kill, release, what-have-you) since cleanup close is so common, but disabling and reenabling gc is usually for timing purposes, and there aren't \*that* many tools that need it.

It doesn't seem all that useful for real time purposes either; sure, it disables gc, but it doesn't disable other forms of implicit non-local code execution that are surprisingly hard to predict (e.g. object destruction, including \_\_del__, for non-cyclic cases is going to depend on whether the stuff being decref-ed is still owned outside the block). Disabling GC means a lot in Java, because \*all* cleanup is GC; in Python, it's just cycle collection, so you're not giving much in the way of guarantees, as the code in question has to be written with a \*lot* of assumptions that Python usually can't support (it's hardly a realtime language).

Seems like if you want a speed up and reliability for this case (and all other context managers intended to be low overhead), it would be better to move parts of contextlib to the C layer (e.g. contextmanager and the classes it's implemented in terms of), so the simple and correct throwaway implementations for arbitrary custom use cases are fast enough; atomicity clearly doesn't actually matter here (it can't be made atomic in any meaningful sense given the lack of thread safety), so any real problem with the provided implementations (assuming they had try/finally added to make them robust) is overhead, not atomicity; the code posted so far would be fine (after adding try/finally) without the speed issues.

**Comment by ncoghlan** [comment]
This is being proposed for the gc module, not contextlib, so I'm not sure how the previous comment applies.

This is just a convenience API for folks already doing this kind of manipulation themselves.

**Comment by rhettinger** [comment]
Assigning this to Lisa for a C implementation, docs, and tests.

**Comment by pablogsal** [comment]
I have prepared a PR in GitHub with an initial implementation of the context manager trying to fulfil the discussed requirements: https://github.com/python/cpython/pull/3980

**Comment by pablogsal** [comment]
I just realize that the link I provided is incorrect. This is the correct one (also is the one that appears in this issue anyway):

https://github.com/python/cpython/pull/4224

**Comment by serhiy-storchaka** [comment]
What is the name of the context manager? gc_disabled, disabled, or Disabled? I see different names in the PR and this confuses me.

**Comment by pablogsal** [comment]
Sorry about that. The context manager is "gc.Disabled()", which I admit is probably a bad name. The documentation is an example of the "equivalent" Python code as stated by Raymond in the first comment but I see now that this will raise confusion. Please, can you indicate what will be a correct name for the context manager so I can update the PR?

**Comment by pablogsal** [comment]
Just to clarify the current situation: At this point, the contextmanager is referred as "disabled" in the C code but is exported as "Disabled" to the garbage collection module. The "gc_disabled" is the Python "equivalent" given by Raymond that is included in the documentation

**Comment by ncoghlan** [comment]
Given the existing "gc.enable", "gc.disable" and "gc.isenabled" APIs, I'd suggest calling this one "gc.ensure_disabled": it ensures the GC is disabled in the with statement body, but only turns it back on at the end if it was previously enabled.

That same name would then be used all the way through the code and documentation.

**Comment by iritkatriel** [comment]
The discussions shows there is no consensus on adding this feature. Marking as pending with a view to closing this (unless someone objects).

**Comment by rhettinger** [comment]
I don't think "non-consensus" is a fair reading.  We made an agreed to commit, but reverted it only because there was a minor issue with it.  The issue is stale because no one has worked on fixing it, but it is still valid.  

The proposal would work fine unless called in multiple threads, so it would need to be documented as not being threadsafe.  It is no more or less powerful that the current way we do it with a `try/finally`.  This would just be a nicer spelling than what we currently have.   This doesn't matter to most users since most people don't use `enable/disable` at all, but it does matter to those are trying to control real-time behavior or those trying to limit reentrancy.   Temporarily turning off GC is one of the few tools available to mitigate the risk of a potentially expensive collection or a collection that triggers reentrant code.  Those users deserve to have the cleaner call afforded by the `with` statement.

**Comment by iritkatriel** [comment]
There were -1s from @1st1 and @serhiy-storchaka, which I interpreted as non-consensus.

@ncoghlan [wrote](https://github.com/python/cpython/issues/75537#issuecomment-1093757790) that there's no point in doing this unless it's done in a thread-safe way.

**Comment by pitrou** [comment]
I do think this would have a value if implemented in a thread-safe way, therefore this issue could probably be kept open.

**Comment by gpshead** [comment]
> the applications for disable/enable tend to be tolerant of an occassional premature enable() call. This is mostly because even with gc disabled, it is very difficult to provide real time guarantees in Python. Disabling GC is a mitigation, not a guarantee. Also, even if another thread reenables GC, it doesn't mean that a collection will happen immediately.

I mostly agreed with the sentiments in that selective quote.  If we were to expand the gc API we need to make these expectations (really "don't expect any of this to reliably ever do what you want") clear in our `gc` module docs.

The difference between `contextlib.chdir` and this for `gc` is that I view a `chdir` as a more common operation - useful for little scripts. While I view the `gc` module as something that probably should've been named `_gc` and is something no user code _should_ ever use. It is an implementation detail, using it is always a code smell.

Adding a context managing variant of `gc.disable()` could be interpreted as blessing the concept as a solution to someones problems. When nothing done with the `gc` module ever is.

That leaves me as -0 on bothering with this even though I'd have found the idea convenient in some tests. Even when I would've reached for it, it was probably still the wrong tool.

**Comment by carljm** [comment]
> While I view the gc module as something that probably should've been named _gc and is something no user code should ever use. It is an implementation detail, using it is always a code smell.

This may be getting OT for this issue, but given how important GC threshold tuning (including, in some cases, disabling automatic GC and only calling `gc.collect()` at certain well-chosen points) can be to Python performance, and how dependent the correct tuning is on the specific characteristics of the workload, and how bad (orders of magnitude too low) the current default thresholds are for many Python workloads, I don't think we can really currently support the idea that use of the `gc` module is always a code smell and no user code should ever use it. I've personally improved the performance of Python workloads by over 30% solely by tuning the GC using APIs in the `gc` module; I'd find it hard to accept that that's something I should not have done.

**Comment by gpshead** [comment]
I'm channeling my ideal world of perfectly spherical theoretical python applications and implementations there...  YouTube did similar gc tweaking.

**Comment by carljm** [comment]
I agree, in the ideal world of perfectly spherical Python, the GC would be so fast and so perfectly self-tuning that no user code would ever have reason to touch it!

**Comment by pitrou** [comment]
> First, there doesn't seem to be a reasonable alternative because the use oflocks has its complications and perils as well. Second, the applications for `disable/enable` tend to be tolerant of an occassional premature `enable()` call. This is mostly because even with gc disabled, it is very difficult to provide real time guarantees in Python. Disabling GC is a mitigation, not a guarantee.

I'm not sure what you're alluding to here. Perhaps we're having a different model of when people might want to selectively disable the GC? My experience is that - apart from uncommon unit tests - the desire to disable the GC comes when particular performance issues are encountered on long-running applications with a large resident working set and very dynamic allocation spikes. Those applications have a high likelihood of being multi-threaded.

**Comment by SakshamKapoor2911** [comment]
PR up at #154999. Adds `gc.ensure_disabled()` context manager using the modern `PyGC_Disable()` C API (which now returns old state).

The original 2018 implementation (PRs #4224/#5456/#5462/#5495/#5496) was reverted due to edge cases during the 3.7 beta window. The C API fix (`PyGC_Disable()` return) and GIL-atomic state management resolve the issues that caused the revert.

@rhettinger @gpshead @ncoghlan — would love your review.
